lib/dencode_context.js

```diff
@@ -0,0 +1,259 @@
+'use strict'; // eslint-disable-line strict
+
+const assert = require('assert');
+const ReedSolomon = require('@ronomon/reed-solomon');
+
+/* Shared, cached, reusable Reed-Solomon context
+ * Thread safe*/
+const _contextCache = new Map();
+
+
+/**
+ * Get Reed Solomon context
+ *
+ * @param {Number} k - Number of data parts
+ * @param {Number} m - Number of coding parts
+ * @return {Object} a Reed-Solomon context
+ * @comment Contexts are cached, never deleted
+*/
+function getECContext(k, m) {
+    assert(typeof k === 'number');
+    assert.ok(k > 0);
+    assert.ok(k <= ReedSolomon.MAX_K);
+    assert(typeof m === 'number');
+    assert.ok(m > 0);
+    assert.ok(m <= ReedSolomon.MAX_M);
+
+    const key = [k, m].join(',');
+    const context = _contextCache.get(key);
+    if (context !== undefined) {
+        return context;
+    }
+
+    // Create an encoding context (can be cached and re-used concurrently):
+    const newContext = ReedSolomon.create(k, m);
+    _contextCache.set(key, newContext);
+    return newContext;
+}
+
+
+/**
+ * Not every stripe sizes are created equal...
+ *
+ * 1/ Stripe size must a multiple of 8:
+ *    required by ReedSolomon lib & Galoid field computations
+ * 2/ Below maximum stream.Readable buffer size
+ *    Refer to actual implementation of computeNewHighWaterMark
+ *    https://github.com/nodejs/node/blob/v6.12.2/lib/_stream_readable.js
+ */
+const MAX_HWM = 0x800000;
+
+/**
+ * Compute a safe stripe size
+ *
+ * @param {Number} k - Number of data parts
+ * @param {Number} m - Number of parity parts
+ * @param {Number} hint - StripeSize hint
+ * @return {Number} actual stripe size to use
+ */
+function safeStripeSize(k, m, hint) {
+    const alignedHint = Math.ceil(hint / 8) * 8;
+    const largest = Math.max(k, m);
```

Comment:
Tiny point: `k` is necessary for object stream but I don't understand why we have to consider `m` here?

Reply:
Indeed, the idea is to limit the maximum read/written size into a stream buffer. Parity streams can only be up to 'hint. Will remove dependency on m then merge.
Thanks for the review :-)

PS: if by any chance you have an idea on how to avoid reallocating each stripe, I'm listening!